File: libs/core/spi.cpp
#include "pxt.h"

// Native SPI master for the `spi` shims (spi.ts). One active bus; the RP2040 SPI
// peripheral (spi0/spi1) is chosen from the SCK pin. On the host VM there's no
// hardware, so calls log via DMESG and reads return 0.

#ifdef PXT_RP2040
#include "hardware/spi.h"
#include "hardware/gpio.h"

// RP2040 pinmux: SPI instance for a GPIO = (gpio / 8) & 1.
//   spi0: GP0-7 & GP16-23 ; spi1: GP8-15 & GP24-29
static spi_inst_t *spi_for(int gpio) {
    return ((gpio >> 3) & 1) ? spi1 : spi0;
}

static spi_inst_t *spiDev;
static int spiFreq = 1000000;
#endif

namespace spi {

//%
void init(int sck, int mosi, int miso) {
#ifdef PXT_RP2040
    spiDev = spi_for(sck);
    spi_init(spiDev, spiFreq);
    gpio_set_function(sck, GPIO_FUNC_SPI);
    gpio_set_function(mosi, GPIO_FUNC_SPI);
    gpio_set_function(miso, GPIO_FUNC_SPI);
#else
    DMESG("spi.init sck=%d mosi=%d miso=%d", sck, mosi, miso);
#endif
}

//%
int write(int value) {
#ifdef PXT_RP2040
    if (!spiDev)
        return 0;
    uint8_t tx = (uint8_t)value, rx = 0;
    spi_write_read_blocking(spiDev, &tx, &rx, 1);
    return rx;
#else
    DMESG("spi.write %d", value);
    return 0;
#endif
}

//%
void transfer(Buffer command, Buffer response) {
#ifdef PXT_RP2040
    if (!spiDev)
        return;
    int len = command ? command->length : (response ? response->length : 0);
    if (len <= 0)
        return;
    const uint8_t *tx = command ? command->data : NULL;
    uint8_t *rx = response ? response->data : NULL;
    if (tx && rx)
        spi_write_read_blocking(spiDev, tx, rx, len);
    else if (tx)
        spi_write_blocking(spiDev, tx, len);
    else
        spi_read_blocking(spiDev, 0, rx, len);
#else
    DMESG("spi.transfer cmd=%d resp=%d", command ? command->length : 0,
          response ? response->length : 0);
#endif
}

//%
void setFrequency(int hz) {
#ifdef PXT_RP2040
    spiFreq = hz;
    if (spiDev)
        spi_set_baudrate(spiDev, hz);
#else
    DMESG("spi.setFrequency %d", hz);
#endif
}

//%
void setMode(int mode) {
#ifdef PXT_RP2040
    if (!spiDev)
        return;
    spi_cpol_t cpol = (mode & 2) ? SPI_CPOL_1 : SPI_CPOL_0;
    spi_cpha_t cpha = (mode & 1) ? SPI_CPHA_1 : SPI_CPHA_0;
    spi_set_format(spiDev, 8, cpol, cpha, SPI_MSB_FIRST);
#else
    DMESG("spi.setMode %d", mode);
#endif
}

} // namespace spi

File: libs/core/spi.ts
/**
 * SPI master communication (one bus).
 */
//% color="#11B5AC" weight=82 icon="" block="SPI"
namespace spi {
    let _inited = false;

    function ensure(): void {
        if (!_inited) {
            // default pins: SCK=GP18, MOSI=GP19, MISO=GP16 (Pico default SPI0)
            init(DigitalPin.P18, DigitalPin.P19, DigitalPin.P16);
        }
    }

    /**
     * Initialize the SPI bus on the given pins (1 MHz, mode 0).
     * The hardware SPI peripheral is chosen automatically from the pins.
     * @param sck the clock pin
     * @param mosi the master-out / data-out pin
     * @param miso the master-in / data-in pin
     */
    //% blockId=spi_init block="SPI init sck %sck|mosi %mosi|miso %miso" weight=100
    export function init(sck: DigitalPin, mosi: DigitalPin, miso: DigitalPin): void {
        __init(sck, mosi, miso);
        _inited = true;
    }

    /**
     * Write a byte to the SPI bus and return the byte read back.
     * @param value the byte to send (0-255)
     */
    //% blockId=spi_write block="SPI write %value" weight=90
    export function write(value: number): number {
        ensure();
        return __write(value & 0xff);
    }

    /**
     * Transfer a buffer over the SPI bus. `command` is sent; received bytes are written
     * into `response` (either may be null).
     */
    //% blockId=spi_transfer block="SPI transfer %command|into %response" weight=85
    export function transfer(command: Buffer, response: Buffer): void {
        ensure();
        __transfer(command, response);
    }

    /**
     * Set the SPI clock frequency in Hz.
     * @param hz frequency, eg: 1000000
     */
    //% blockId=spi_freq block="SPI set frequency %hz" weight=80
    export function setFrequency(hz: number): void {
        ensure();
        __setFrequency(hz);
    }

    /**
     * Set the SPI mode (0-3): clock polarity/phase.
     * @param mode 0,1,2 or 3
     */
    //% blockId=spi_mode block="SPI set mode %mode" mode.min=0 mode.max=3 weight=75
    export function setMode(mode: number): void {
        ensure();
        __setMode(mode & 3);
    }
}

declare namespace spi {
    //% shim=spi::init
    function __init(sck: DigitalPin, mosi: DigitalPin, miso: DigitalPin): void;
    //% shim=spi::write
    function __write(value: int32): int32;
    //% shim=spi::transfer
    function __transfer(command: Buffer, response: Buffer): void;
    //% shim=spi::setFrequency
    function __setFrequency(hz: int32): void;
    //% shim=spi::setMode
    function __setMode(mode: int32): void;
}
